server/src/test/java/org/elasticsearch/gateway/RecoveryFromGatewayIT.java

```diff
@@ -567,4 +571,60 @@ public Settings onNodeStopped(String nodeName) throws Exception {
+
+        // This test pertains to a race condition in which deleting a shard concurrently with a TransportNodesListGatewayStartedShards
+        // request could resurrect the shard's metadata folder after it was deleted. Here we try and recreate the race, but it is quite
+        // delicate so this test does not always fail. Experimentation showed that setting the thread counts as below would yield a failure
+        // after a reasonable number of iterations: running repeatedly with -Dtests.iters=1000 saw 6 failures out of 10 runs.
+        //
+        // NB this experiment was run with
+        // @TestLogging("org.elasticsearch.env.NodeEnvironment:TRACE,org.elasticsearch.gateway.MetaDataStateFormat:TRACE," +
+        //     "org.elasticsearch.gateway.TransportNodesListGatewayStartedShards:TRACE,org.elasticsearch.index.shard.IndexShard:TRACE")
+        // but with less verbose logging the failures seem rarer.
+
+        final String nodeName = internalCluster().startNode();
+        DiscoveryNode node = internalCluster().getInstance(ClusterService.class, nodeName).localNode();
+
+        assertAcked(prepareCreate("test").setSettings(Settings.builder()
+            .put(SETTING_NUMBER_OF_SHARDS, 1).put(SETTING_NUMBER_OF_REPLICAS, 0)));
+        final ShardId shardId = new ShardId(resolveIndex("test"), 0);
+
+        final int listingThreadCount = 2;
+        final int deletingThreadCount = 2;
+
+        final CountDownLatch countDownLatch = new CountDownLatch(listingThreadCount + deletingThreadCount);
+
+        Thread threads[] = new Thread[listingThreadCount + deletingThreadCount];
+        for (int threadIndex = 0; threadIndex < listingThreadCount + deletingThreadCount; threadIndex++) {
+            final boolean isListingThread = threadIndex < listingThreadCount;
+            threads[threadIndex] = new Thread(() -> {
+                try {
+                    countDownLatch.countDown();
+                    countDownLatch.await();
+
+                    if (isListingThread) {
+                        internalCluster().getInstance(TransportNodesListGatewayStartedShards.class)
+                            .execute(new TransportNodesListGatewayStartedShards.Request(shardId, new DiscoveryNode[]{node}))
+                            .get();
+                    } else {
+                        assertAcked(client().admin().indices().prepareDelete("test"));
+                    }
+                } catch (InterruptedException | ExecutionException | IndexNotFoundException ignored) {
+                    // don't care if this fails
+                }
+            }, (isListingThread ? "Listing" : "Deleting") + "[" + threadIndex + "]");
+        }
+
+        for (final Thread listingThread : threads) {
+            listingThread.start();
+        }
+
+        // Deleting an index asserts that it really is gone from disk, so no other assertions are necessary here.
```

Comment:
I think it's still good to have this assertion here? it's clearer and we don't rely on other code that may change without taking this test into account.

Reply:
Good point, I pushed 48f6d46d0a7